[clang][DependencyScanning] Track modules that resolve from "shared" locations

cyndyishida opened the pull request:
That patch tracks whether all the file & module dependencies of a module resolve to a sysroot location. This information will later be queried by build systems for determining where to store the accompanying pcms.

cyndyishida committed:
[clang][DependencyScanning] Track modules that resolve from sysroot.

That patch tracks whether all the file & module dependencies of a module
resolve to a sysroot location. This information will later be queried by
build systems for determining where to store the accompanying pcms.

cyndyishida commented:
> but I'm concerned that this does not guarantee the resulting command line only refers to inputs in the sysroot. Imagine that we run this without the header search and VFS pruning optimizations - the command line will likely refer to unused project-specific source/build directories.

I would think that if `ModuleDepCollector` only found dependency inputs that resolve to sysroot locations, the command line for building the specific module wouldn't matter. Of course, I would also expect that `Precompile module` invocations don't point to paths they don't need, but wouldn't be required for tracking whether a module is in sysroot.  Is there a concern that the command lines (e.g. header search paths with different files populated) between scanning and compiling could result in different input files being resolved?

> One way of doing this is going through the *Options classes that contribute to CompilerInvocation and adding an API that allows for iteration through all input/output paths. 

Do you have an intuition of whether this would be noticeably expensive for the scanner?

cyndyishida committed:
Add explicit search paths into SDK as they aren't passed with non-darwin targets

Bigcheese commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -739,6 +748,11 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
   MDC.ScanInstance.getASTReader()->visitInputFileInfos(
       *MF, /*IncludeSystem=*/true,
       [&](const serialization::InputFileInfo &IFI, bool IsSystem) {
+        auto FullFilePath = ASTReader::ResolveImportedPath(
+            PathBuf, IFI.UnresolvedImportedFilename, MF->BaseDirectory);
+        if (MD.IsInSysroot)
+          MD.IsInSysroot = FullFilePath->starts_with(CurrSysroot);
+        PathBuf.resize_for_overwrite(256);
```
I think it would be good to also gate the call to `ResolveImportedPath` on `IsInSysroot`. There is a bit of overhead to calling that.

Bigcheese commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -843,7 +864,7 @@ void ModuleDepCollectorPP::addModuleDep(
         !MDC.isPrebuiltModule(Import)) {
```
I don't think this handles non-sysroot prebuilt modules. I think that if you have a PCH that depends on a non-sysroot module then this doesn't get tracked.

cyndyishida replied:
In hindsight, I wish I made this a stacked diff against the real repo.

I have an implementation that addresses this in https://github.com/cyndyishida/llvm-project/commit/272b49206c51c4cc8879c2b539f199c5548fb47a I'd like to merge the current PR and then open a new PR against main for it. But I am open to other suggestions to make the review easier. 

That implementation adds some overhead for two reasons.
1. It deserializes out more content from prebuilt pcm files. 
2. When a module path is fully realized, it traverses down into the leaf dependency and updates whether they are in the sysroot. 

I measured this overhead on a large-ish objc project that uses pch's and toy examples against the macos sdk. I did not observe a noticeable difference.

Bigcheese replied:
I think that's fine as long as the current patch says that pre built modules means it's not in sysroot, as I think that's a small addition and keeps it conservatively correct.

cyndyishida replied:
Added a test for this, modified from the working patch. https://github.com/llvm/llvm-project/pull/130634/files#diff-06c3b647f63ed03628260e6d6c1e679990852e445d03dd60d4466248c1f32a5a

Bigcheese commented on clang/include/clang/Tooling/DependencyScanning/ModuleDepCollector.h:
```diff
@@ -114,6 +114,10 @@ struct ModuleDeps {
   /// Whether this is a "system" module.
   bool IsSystem;
 
+  /// Whether this is a module where it's dependencies resolve within the
+  /// sysroot.
```
I think this should say that the external name resolves within the sysroot to be clear if it's the virtual or external name. Also maybe something about why this exists like "this is used to identify modules that are unlikely to change during normal development".

cyndyishida committed:
Address review comments except for prebuilt module dependencies

benlangmuir commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -739,6 +748,12 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
+        if (MD.IsInSysroot) {
+          auto FullFilePath = ASTReader::ResolveImportedPath(
+              PathBuf, IFI.UnresolvedImportedFilename, MF->BaseDirectory);
+          MD.IsInSysroot = FullFilePath->starts_with(CurrSysroot);
```
Minor nit: string prefix is not correct for paths, it also needs to check that the next character in `FullFilePath` is a separator.  I'm surprised I can't find an existing utility for this in Path.h, but maybe we should add one.

cyndyishida replied:
There is one private to `Path.cpp` I could expose that?

benlangmuir replied:
Do you mean `starts_with`? That one looks like it has the same issue 😬 In fact I see in the `replace_path_prefix` documentation it claims the following behaviour
```
/oldfoo, /old, /new => /oldfoo
```

but the unit test for this function actually has an equivalent case and the behaviour is different `/oldnew/foo, /old, /new => /newnew/foo` specifically because `starts_with` is wrong.

cyndyishida replied:
There's also a swift version that attempts to do the same. https://github.com/swiftlang/swift/blob/b4a2908fb5d7ae5b19d1f4a7f0ec9636f7c729c9/lib/Basic/StringExtras.cpp#L1459
It isolates & moves forward by separator, so I can try to apply this here (and double check it resolves the examples you point out).

cyndyishida replied:
I modified the test case to handle a case like you described. https://github.com/llvm/llvm-project/pull/130634/files#diff-988370ab7ec8ac1b26c1fa07bda79906b52d097324dac9a9d04774debf330a6cR70

benlangmuir commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -835,6 +850,13 @@ void ModuleDepCollectorPP::addAllSubmoduleDeps(
   });
 }
 
+void ModuleDepCollectorPP::addClangModule(const Module *M, const ModuleID ID,
```
How about something like "addOneModuleDep"?  If find it hard to guess from the name how this is different from e.g. `addModuleDep` which also has a similar signature.

benlangmuir reviewed (commented):
I have a similar concern to @jansvoboda11 that we probably need to check the invocation paths as well.  

> I would think that if ModuleDepCollector only found dependency inputs that resolve to sysroot locations, the command line for building the specific module wouldn't matter.

One concern I have is if we have user search paths, the contents of those paths can change after scanning in a way that silently affects the compilation. Another more theoretical issue is that in general the set of module input files has no guarantee of being complete -- it's derived from the source manager and we don't necessarily have a mechanism to ensure every file that impacted the module was captured.

Bigcheese reviewed (commented):
Overall I think this looks good. I'm not sure I share the same concerns. The situation this covers is how likely is it that the build system is going to determine this needs to be rebuilt, and that happens (in most cases) purely based on the reported dependencies. So negative deps or unlisted deps don't change anything here.

cyndyishida commented:
> I have a similar concern to @jansvoboda11 that we probably need to check the invocation paths as well.

> The situation this covers is how likely is it that the build system is going to determine this needs to be rebuilt, and that happens (in most cases) purely based on the reported dependencies.

I can add the checks for inputs through the compiler invocations, with the hope that it will allow for more pcm sharing across build plans (e.g. projects) depending on the same modules. I am a little worried that a more conservative check would mean fewer modules that otherwise could safely be in a shared location don't end up there due to other limitations (e.g. we don't prune out an option that is safe to), but I don't have a sense if that is a practical concern.

cyndyishida committed:
Update IsInSysroot -> IsShareable

* We need to track more than just sysroot, so start with sysroot and
  resource directory as a starting point, but allow the ability to
capture more as they are discovered in the future.

* Check a subset of the compiler invocation of module compilations for
  determining if a module can be shared.
* When a module has a prebuilt module dependency, consider is not
  shareable for now.

benlangmuir commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -739,6 +795,12 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
+          auto FullFilePath = ASTReader::ResolveImportedPath(
+              PathBuf, IFI.UnresolvedImportedFilename, MF->BaseDirectory);
+          MD.IsShareable = isPathInSharedDir(SharedDirs, *FullFilePath);
+          PathBuf.resize_for_overwrite(256);
```
Why are you resizing the string here? `ResolveImportedPath` will `clear()` it before using it anyway.

cyndyishida replied:
I thought I was running into the assertion failure in `ResolveImportedPath`, but I can't reproduce it now.

benlangmuir commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -157,6 +157,32 @@ static void optimizeCWD(CowCompilerInvocation &BuildInvocation, StringRef CWD) {
+static bool areOptionsInSharedDir(CowCompilerInvocation &BuildInvocation,
+                                  const ArrayRef<StringRef> SharedDirs) {
+  const auto &HSOpts = BuildInvocation.getHeaderSearchOpts();
+  if (!isPathInSharedDir(SharedDirs, HSOpts.Sysroot))
```
Should the first two checks be asserts since we derive sharable directories from sysroot and resource dir?  Or is there a case where they will differ?

cyndyishida replied:
I can't think of this happening intentionally.

cyndyishida committed:
Address review comments

Bigcheese reviewed (approved):
lgtm now, just a request to change the name.

cyndyishida commented on clang/include/clang/Tooling/DependencyScanning/ModuleDepCollector.h:
```diff
@@ -114,6 +114,14 @@ struct ModuleDeps {
+  ///
+  /// External paths, as opposed to virtual file paths, are always used
+  /// for computing this value.
+  bool IsShareable;
```
> I'm not sure I like this name, as it implies if this is false they aren't sharable.

Why do you find this implication inaccurate? Like if the compiler cannot conclude the pcm is shareable, it is functionally not-shareable.

cyndyishida replied:
Settled on `IsInStableDirectories`

cyndyishida committed:
IsShareable -> IsInStableDirectories
